### PR TITLE
MSE regression

### LINKED ISSUE #52: VarRecorder should be able to record connection variables
Because of ambiguity when pointing at ``Layer`` objects, maybe using a derived class/flag

### PR DESCRIPTION
**Pull request by neworderofjamie**
This PR extends the existing mechanism in the EventProp compiler used to support 'per-timestep' cross-entropy loss to also supports mean-squared error loss for regression tasks. Using this, the pattern recognition task (previously used as an e-prop example) has been implemented as an EventProp example (encountering #101 and #102 in the process)

Also, this PR fixes a number of corner cases involving recording of non-batched variables in batched simulations and implements ``ConnVarRecorder`` for recording from ``Connection`` objects.

Fixes #52

### CODE REVIEW

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 1]
```diff
@@ -502,7 +510,7 @@ def build_neuron_model(self, pop: Population, model: NeuronModel,
                     """)
 
                 # If we want to calculate per-timestep loss
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:512]:
```suggestion
                # If we want to calculate mean-squared error or per-timestep loss
```

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 2]
```diff
@@ -511,120 +519,154 @@ def build_neuron_model(self, pop: Population, model: NeuronModel,
...
+                            # Add custom update to reset state
+                            compile_state.add_neuron_reset_vars(
+                                pop, reset_vars, False, True)
+                            
+                        # Otherwise, unsupported readout type
+                        else:
+                            raise NotImplementedError(
+                                f"EventProp compiler with CategoricalCrossEntropy loss doesn't support "
+                                f"{type(pop.neuron.readout).__name__} readouts")
+                    if mse_loss:
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:556]:
```suggestion
                    elif mse_loss:
```

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 3]
```diff
@@ -511,120 +519,154 @@ def build_neuron_model(self, pop: Population, model: NeuronModel,
...
+                            compile_state.add_neuron_reset_vars(
+                                pop, 
+                                [("VMaxTimeBack", "scalar", "VMaxTime")] + model_copy.reset_vars, 
+                                False, False)
+                        # Otherwise, unsupported readout type
+                        else:
+                            raise NotImplementedError(
+                                f"EventProp compiler doesn't support "
+                                f"{type(pop.neuron.readout).__name__} readouts")
+                    if mse_loss:
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:666]:
```suggestion
                    elif mse_loss:
```

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 4]
```diff
@@ -463,13 +464,20 @@ def build_neuron_model(self, pop: Population, model: NeuronModel,
...
-                               VarAccess.READ_ONLY_SHARED_NEURON)
-
+            if sce_loss:
+                model_copy.add_var("YTrue", "uint8_t", 0, 
+                                   VarAccess.READ_ONLY_SHARED_NEURON)
+
+                # Add second variable to hold the true label for the backward pass
+                model_copy.add_var("YTrueBack", "uint8_t", 0, 
+                                   VarAccess.READ_ONLY_SHARED_NEURON)
+            if mse_loss:
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:474]:
```suggestion
            elif mse_loss:
```

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 5]
```diff
@@ -463,13 +464,20 @@ def build_neuron_model(self, pop: Population, model: NeuronModel,
             # Add variable, shared across neurons to hold true label for batch
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:464]:
```suggestion
```

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 6]
```diff
@@ -463,13 +464,20 @@ def build_neuron_model(self, pop: Population, model: NeuronModel,
...
             # it will add unwanted code to end of neuron but we do want this
-            model_copy.add_var("YTrue", "uint8_t", 0, 
-                               VarAccess.READ_ONLY_SHARED_NEURON)
-
-            # Add second variable to hold the true label for the backward pass
-            model_copy.add_var("YTrueBack", "uint8_t", 0, 
-                               VarAccess.READ_ONLY_SHARED_NEURON)
-
+            if sce_loss:
+                model_copy.add_var("YTrue", "uint8_t", 0,
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:468]:
```suggestion
                # Add variable, shared across neurons to hold true label for batch
                model_copy.add_var("YTrue", "uint8_t", 0, 
```

File: examples/event_prop/pattern_recognition.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+'''
+for (pop,pypop) in compiled_net.neuron_populations.items():
+    print(pypop.model.get_sim_code())
+    for var in pypop.model.get_vars():
+        print(var.name)
+    for para in pypop.model.get_params():
+        print(para.name)
+    for dp in pypop.model.get_derived_params():
+        print(dp.name)
+'''
```
Comment (neworderofjamie) [inline examples/event_prop/pattern_recognition.py:107]:
```suggestion
```

File: examples/event_prop/pattern_recognition.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+        else:
+            return alpha
+
+    # Evaluate model on numpy dataset
+    start_time = perf_counter()
+    callbacks = ["batch_progress_bar", 
+                 VarRecorder(output, "v", key="output_v"),
+                 VarRecorder(output, genn_var="LambdaV", key="output_lambdav"),
+                 VarRecorder(output, genn_var="LambdaI", key="output_lambdai"),
+                 SynVarRecorder(h2o, genn_var="dw", key="h2o_dw"),
```
Comment (neworderofjamie) [inline examples/event_prop/pattern_recognition.py:122]: Not sure if we want to record all internal state in the example

File: ml_genn/ml_genn/callbacks/spike_recorder.py [hunk 9]
```diff
@@ -31,6 +31,7 @@ def __init__(self, pop: PopulationType, key=None,
 
         # Stash key and whether we're recording spikes or spike-like events
         self.key = key
+
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/callbacks/spike_recorder.py:34]:
```suggestion
```

File: ml_genn/ml_genn/callbacks/var_recorder.py [hunk 10]
```diff
@@ -105,7 +105,7 @@ def on_timestep_end(self, timestep: int):
         if np.any(self._batch_mask):
             # Copy variable from device
             pop = self._compiled_network.neuron_populations[self._pop]
-            pop.pull_var_from_device(self._var)
+            pop.vars[self._var].pull_from_device()
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/callbacks/var_recorder.py:108]: Yey! Clearly **some** of the warnings have their desired effect 😄

**Review by neworderofjamie — COMMENTED** [review]
All looks good aside from some very minor formatting I've suggested

**Review by tnowotny — APPROVED** [review]
I think this is ready to be merged as far as I can see?

File: ml_genn/ml_genn/callbacks/__init__.py [hunk 11]
```diff
@@ -1,19 +1,20 @@
...
-__all__ = ["Callback", "Checkpoint", "CustomUpdateOnBatchBegin",
-           "CustomUpdateOnBatchEnd", "CustomUpdateOnTimestepBegin",
-           "CustomUpdateOnTimestepEnd", "OptimiserParamSchedule",
-           "BatchProgressBar", "SpikeRecorder", "VarRecorder",
-           "default_callbacks"]
+__all__ = ["Callback", "Checkpoint", "ConnVarRecorder", 
+           "CustomUpdateOnBatchBegin", "CustomUpdateOnBatchEnd", 
+           "CustomUpdateOnTimestepBegin", "CustomUpdateOnTimestepEnd", 
+           "OptimiserParamSchedule", "BatchProgressBar", "SpikeRecorder",
+           "VarRecorder", "ConnVarRecorder", "default_callbacks"]
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/callbacks/__init__.py:20]:
```suggestion
           "VarRecorder", "default_callbacks"]
```
bad merge here

File: ml_genn/ml_genn/losses/mean_square_error.py [hunk 12]
```diff
@@ -18,8 +18,8 @@ def add_to_neuron(self, model: NeuronModel, shape,
         model.append_sim_code(
             f"""
             const unsigned int timestep = (int)round(t / dt);
-            const unsigned int index = (timestep * num_batch * num_neurons)
-                                       + (batch * num_neurons) + id;
+            const unsigned int index = (batch * {example_stimesteps} * num_neurons)
```
Comment (neworderofjamie) [inline ml_genn/ml_genn/losses/mean_square_error.py:21]:
```suggestion
            const unsigned int index = (batch * {example_timesteps} * num_neurons)
```

**Review by neworderofjamie — COMMENTED** [review]
couple of typos but all looks good aside from a bit of documentation on ``ConnVarRecorder``

File: ml_genn/ml_genn/losses/mean_square_error.py (hunk 12 above)
Comment (tnowotny) [reply ml_genn/ml_genn/losses/mean_square_error.py:21]: remarkable how I could have accidentally inserted "s"!

**Review by tnowotny — APPROVED** [review]
I think this looks good to go.

**Pull request merged**